### PR TITLE
Simplify implementation of Model.compute and improve caching

### LINKED ISSUE #1671: caching in `models.interface`
Might be included in another issue already?
TODOs by @sdrave are
https://github.com/pymor/pymor/blob/main/src/pymor/models/interface.py#L373
https://github.com/pymor/pymor/blob/main/src/pymor/models/interface.py#L394
https://github.com/pymor/pymor/blob/main/src/pymor/models/interface.py#L405
https://github.com/pymor/pymor/blob/main/src/pymor/models/interface.py#L414

### PR DESCRIPTION
**Pull request by sdrave**
This PR is a complete redesign and simplification of the implementation of `Model.compute`. Caching is significantly improved: now all computable quantities like the model's output or parameter sensitivities are cacheable, not only the solution. Subsequent calls to `compute` benefit from caching of previously computed quantities that are needed as intermediate results.

- `_compute_solution`, `_compute_output`, etc. are removed. All implementations are now located in a single `_compute` method, which can use default implementation by using `super`.
- `_compute_allowed_kwargs` is replaced by a public `computable_quantities` property which lists all allowed arguments for `compute`.
- `_compute_required_quantities` can be called in `_compute` to request additional quantities that are needed to compute a given quantity.
- The new `data` parameter of `compute` can be used to pass previously computed quantities as intermediate values in case caching is not enabled.
- `CacheableObject` gathers the new low-level methods `get_cached_value` and `set_cached_values` for caching values which are not tied to a method call.

Depends on #2278.
Fixes #1671.

### CODE REVIEW

File: src/pymor/models/iosys.py [hunk 1]
```diff
@@ -675,55 +675,59 @@ def to_abcde_files(self, files_basename, mu=None):
...
+            if compute_output:
+                if data_output_extra:
+                    data['output'] = np.vstack((data['output'], data_output_extra))
+                if len(data['output']) < i + 1:
+                    data['output'] = data['output'][:i + 1]
+
+            if compute_solution:
+                quantities.remove('solution')
+            if compute_output:
+                quantities.remove('output')
```
Comment (HenKlei) [inline src/pymor/models/iosys.py:728]:
```suggestion
            if compute_solution:
                data['solution'] = self.solution_space.empty(reserve=n)
                time_instances = []
                for i, (x, t) in enumerate(iterator):
                    data['solution'].append(x)
                    time_instances.append(t)
                quantities.remove('solution')
            if compute_output:
                D = LinearInputOperator(self.D)
                data['output'] = np.empty((n, self.dim_output))
                data_output_extra = []
                for i, (x, t) in enumerate(zip(data['solution'], time_instances)):
                    y = self.C.apply(x, mu=mu).to_numpy() + D.as_range_array(mu=mu.with_(t=t)).to_numpy()
                    if i < n:
                        data['output'][i] = y
                    else:
                        data_output_extra.append(y)
                if data_output_extra:
                    data['output'] = np.vstack((data['output'], data_output_extra))
                if len(data['output']) < i + 1:
                    data['output'] = data['output'][:i + 1]
                quantities.remove('output')
```

Just as an idea for how to get rid of all the `if compute_...`.

**Review by HenKlei — COMMENTED** [review]
No major comments from my side, looks all reliable I'd say.

File: src/pymor/models/basic.py [hunk 2]
```diff
@@ -89,59 +91,45 @@ def __str__(self):
...
+
+            jacobian = self.output_functional.jacobian(data['solution'], mu)
             assert jacobian.linear
             dual_solutions = self.operator.range.empty()
             for d in range(self.output_functional.range.dim):
                 dual_problem = self.with_(operator=self.operator.H,
                                           rhs=jacobian.H.as_range_array(mu)[d])
                 dual_solutions.append(dual_problem.solve(mu))
-            sensitivities = {}
+            sensitivites = {}
```
[the remark this answers is not in the document]
Comment (sdrave) [reply src/pymor/models/basic.py:120]: Huh, no idea how that happend ..

File: src/pymor/models/interface.py [hunk 3]
```diff
@@ -601,6 +451,100 @@ def visualize(self, U, **kwargs):
...
+        `solution_error_estimate` and `output_error_estimate` quantities.
+        The implementations for `output` and `output_d_mu` require the model to posses
+        an `output_functional` attribute, which is applied to the solution in order to
+        obtain the output. `solution_error_estimate` and `output_error_estimate` defer
+        the error estimation to the model's :attr:`error_estimator`.
+
+        In case a requested quantity depends on another quantities, implementations should
+        call ::
+
+            self._compute_required_quantities({'quantity_a', 'quantity_b'}, data, mu)
```
[the remark this answers is not in the document]
Comment (sdrave) [reply src/pymor/models/interface.py:478]: That would be the case, yes. However, I don't see anyone doing that.

File: src/pymor/models/iosys.py (hunk 1 above)
Comment (sdrave) [reply src/pymor/models/iosys.py:728]: Maybe I misunderstand your suggestion, but the point of the original code is that when you are only interested in the output you don't need to keep the entire state space trajectory in memory ..

Comment (HenKlei) [reply src/pymor/models/iosys.py:728]: Ah yes, of course. You are right!

File: src/pymor/models/interface.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HenKlei) [reply src/pymor/models/interface.py:478]: Hopefully not ;-)

**Review by pmli — APPROVED** [review]

**Pull request merged**